### PR TITLE
7.x filter html

### PR DESCRIPTION
**Pull request by MorganDawe**
**JIRA Ticket**: [ISLANDORA-2017](https://jira.duraspace.org/browse/ISLANDORA-2017)

# What does this Pull Request do?

When displaying a metadata field that is both truncated to a max length AND contains basic HTML tags, the 'Show more' and 'Show less' toggle can break the display. While modules that implement this functionality (such as islandora_solr_metadata) performs a check_markup() call in the template file (which uses the 'islandora_solr_metadata_filtered_html' text format for correctness), we should also be preforming a check like this on the raw truncated value of the field, as well as the full value of the field. This pull request will call the filter modules '_filter_htmlcorrector()' function to ensure metadata tags are correctly closed.

# What's new?
This pull request will add two calls to '_filter_htmlcorrector()' to be sure field content that contains markup will properly be closed, prior to attempting to link the field value in question (if configured to do so) or appending the show more/show less toggle.

# How should this be tested?

* Add metadata field content to an ingested objects field (such as 'description') that contains basic HTML
* Truncate the whole field to a max length, so that the max length falls between an opening and closing HTML tag
* View the field in a search result, or within the islandora_solr_metadata display

# Additional Notes:
It may also be worth noting, the islandora_solr_metadata module template is wrapping description data in a `<p>` tag, which should not semantically contain HTML tags [see here](https://github.com/Islandora/islandora_solr_metadata/blob/7.x/theme/islandora-solr-metadata-description.tpl.php#L24). While metadata fields do not typically include HTML tags, it is entirely possible they will.

* Could this change impact execution of existing code?
I do not _believe_ this change will change the execution of existing code, however it could change the fields display if said field already contains basic html, and the islandora_solr_metadata_filtered_html text formatter is configured to allow it.

# Interested parties
Taging @nhart  @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by DiegoPino** [comment]
Hey @MorganDawe thanks, this was something required! Wonder if we could discuss fixing the issue from its root instead of filtering even more? 'Show more' and 'Show less' are hackish-ish inserted as HTML into the render array, how do you feel about making that a proper theme() piece and instead of displaying/rendering it right away passing it along as an expanded /not rendered render array so others can really extend and change that? Also, an extra dependency could be avoided that way

**Comment by MorganDawe** [comment]
@DiegoPino I am all for refactoring this into a theming piece, but i would like to maintain much of the leg work in preparing the vars for theming in the [islandora_solr_truncate_field_display()](https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L488) function, thus maintaining compatibility with with other modules that may already implement it. I am thinking migrating the functionality from right around [here](https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L520) down into the new template_preprocess function. This would make it easier for any additional modules that implement this new template preprocessing hook i think.

**Comment by DiegoPino** [comment]
@MorganDawe cool. Yeah, there is a lot of useful code there and for sure, the logic should be maintained and honored. Just thinking we should get rid of things like https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L551 in favor of that theming proposal/implementation you just proposed. I know filtering is always a good solution, but also a solution that can/could have so many exceptions, fire, and smoke, and also makes things always slow. Let's talk about this during today's @Islandora/7-x-1-x-committers call.  Thanks!

**Comment by adam-vessey** [comment]
PR description indicates the show more/less behaviour, implying the use of "Limit Length of the whole field" business; however, this same issue presents itself with "Limit length of each separate value". It looks like this could be addressed by adding a `_filter_htmlcorrector()` call around [where the separate values are handled](https://github.com/Islandora/islandora_solr_search/blob/00557f71b01e996f33e0bdc381ace999e8591959/includes/utilities.inc#L371).

**Comment by DiegoPino** [comment]
@MorganDawe ping me when you are ready/want to test this. Thanks

**Comment by MorganDawe** [comment]
Big thumbs up to @adam-vessey for doing 99% of the heavy lifting in the new IslandoraSolrFieldTruncation Class that is now being implemented here. The goal of the class is to properly truncate metadata fields that may contain markup. An issue may arise with markup that has a number of attributes and properties when preforming truncation previously (say, truncating text in the middle of an html elements class name). Processing DOMElements should ensure this does not happen. Small refactor to come.

**Comment by DiegoPino** [comment]
@adam-vessey++ really good work. 👍

**Comment by MorganDawe** [comment]
@DiegoPino We are still preforming an internal review of this code, but appears functional if you have time to give it a shot. Thank you very much.

File: includes/utilities.inc
Comment (qadan) [inline includes/utilities.inc]: Should we perhaps be casting to string at this point? Didn't want to do the level of digging necessary to figure out if there are any cases in here or modules out there that might be expecting a string here.

File: includes/field_truncation.inc
Comment (qadan) [inline includes/field_truncation.inc]: Param type hint should be `int`, no?

File: includes/field_truncation.inc
Comment (qadan) [inline includes/field_truncation.inc]: Similarly, param type hint appears like it should be `bool`?

File: includes/field_truncation.inc
Comment (qadan) [inline includes/field_truncation.inc]: Type hint `int`?

File: includes/field_truncation.inc
Comment (qadan) [inline includes/field_truncation.inc]: Type hint `int`?

**Review by qadan — COMMENTED** [review]
Worry about the value generation (used to be providing a string explicitly, now providing an object that can be cast to a string. Also some type hinting stuff

**Comment by DiegoPino** [comment]
@MorganDawe once internal review is done will be happy to test. Good luck

**Comment by MorganDawe** [comment]
@DiegoPino This work is rounding the corner of our QA team now, so you should be good to test when your available. Thanks again.

**Comment by DiegoPino** [comment]
@MorganDawe ok, have some time tonight, will be done by tomorrow AM, thanks

**Comment by MorganDawe** [comment]
@DiegoPino If you get a chance to take a look, i would value your feedback on this pull. Thank you.

**Comment by DiegoPino** [comment]
@MorganDawe almost there. Looks good. Need to test a bit more. That class is kinda neat!

File: includes/utilities.inc
Comment (DiegoPino) [inline includes/utilities.inc:523]: What about using actuall &hellip; == `&hellip;` instead of ...

File: includes/field_truncation.inc
Comment (DiegoPino) [inline includes/field_truncation.inc:50]: Wonder if checking for < and > could lead to misleading "HTML" is there. What about a regular expression? '/(<\/?([\w+]+)[^>]*>)?([^<>]*)/', where you match tags and text independently? What happens if we actually get < and > but in this order ><...will XML/Dom Parser not fail? Sorry, don´t want to be complicated, the class is neat and well written, and Dom is more trustable than a regular expression (which could be faster...depending really), but passing a non valid HTML piece to DOM can make things break? Maybe tiny? Maybe just dilousional.

**Review by DiegoPino — COMMENTED** [review]
Just some comments. The current implementation of this class makes kinda sure it will not break, but the class could break? if passed something differently. So added a few notes.

File: includes/field_truncation.inc
Comment (MorganDawe) [reply includes/field_truncation.inc:50]: I am a little hesitant to implement a reg ex here, feels like there may be more of a margin of error. As well, according to the [preg_match()](http://php.net/manual/en/function.preg-match.php) page it looks like it would be faster to use a strpos(). I have added to local code a fallback to use $dom->loadHTML() if the loadXML() fails in the case strings like the ones you mentioned (><, for example) are in the text.

File: includes/utilities.inc
Comment (MorganDawe) [reply includes/utilities.inc:523]: the call to truncate_utf8() will add the ellipsis in the truncateText member function as its fourth parameter. This section just counts the characters to substitute from the max length. Not sure if there is a point in useing `&hellip;` here. I  could be crazy, just does not appear to work here.

**Comment by MorganDawe** [comment]
Any more comments or concerns on this?

**Review by DiegoPino — APPROVED** [review]
Looks good on this side but i can't deep test this until monday, so please feel to do so before that.

**Comment by jordandukart** [comment]
Bump @DiegoPino. Perhaps @dannylamb can look at this seeing as it's been open for > a month.

**Comment by DiegoPino** [comment]
@jordandukart this was good to go.Sorry, got mixed up. All concerns addressed. Merging

**Comment by DiegoPino** [comment]
@jordandukart @MorganDawe 👀 merged.

**Pull request merged**
